### ISSUE TITLE
"No results found" shows up on every empty line and on every token including keywords

### ISSUE THREAD

**Issue by macraig** [sourcegraph/sourcegraph#33558]
@mrnugget's findings:

1. We only show the spinner after 600ms. That can lead to some rather tricky behaviour: if the first request is slow and doesn't find results, you'll see a spinner and then the hover disappears. If you then hover over the token again, the response from the previous request is cached, which means it'll not cross the 600ms, which means you won't see the spinner or a hover tooltip again. https://github.com/sourcegraph/sourcegraph/blob/93eebb8b847a92675b181fd34cef8501f0b2ea06/client/codeintellify/src/hoverifier.ts#L345-L346 
2. Our code contains some (really old) logic to display "nothing found" when there is no result: https://github.com/sourcegraph/sourcegraph/blob/93eebb8b847a92675b181fd34cef8501f0b2ea06/client%2Fshared%2Fsrc%2Fhover%2FHoverOverlayContents%2FHoverOverlayContents.tsx#L54-L59
3. That code is never triggered, because the hover overlay hides the hover if there is no cursor position provided: https://github.com/sourcegraph/sourcegraph/blob/93eebb8b847a92675b181fd34cef8501f0b2ea06/client%2Fshared%2Fsrc%2Fhover%2FHoverOverlay.tsx#L63-L74
4. On the other end of things we have some code that also decides when to render a hover tooltip and when not. And if it doesn't have content for the hover (but maybe actions?) then it doesn't return anything, which means it doesn't return a position, which leads to code in (2) not being shown, due to (3). https://github.com/sourcegraph/sourcegraph/blob/93eebb8b847a92675b181fd34cef8501f0b2ea06/client%2Fcodeintellify%2Fsrc%2Fhoverifier.ts#L308-L319
5. I changed the logic in (4) to always return the position, even if there is no hover content and now we can see the "No results found" content:

https://user-images.githubusercontent.com/1185253/157459009-0211fb15-1337-4bd2-bd2e-4ed6100edda6.mp4

6. The problem: we don't distinguish between "no result found and we'll _never_ find a result for a token like this (whitespace, parentheses, ...)" and "no result found, but maybe with precise", which leads to the "no results found" showing up on every empty line and on every token, including keywords.

**Comment by sourcegraph-bot-2** [comment]
Heads up @macraig - the "team/code-intelligence" label was applied to this issue.

**Comment by chrismwendt** [comment]
It's not clear to me that there's a real issue here. As far as I can tell, it is intended behavior for a spinner to appear then disappear once the response arrives with no results.

**Comment by chrismwendt** [comment]
UX/design question: when there is no result, should the hover popover tell the user?

**Comment by mrnugget** [comment]
The current title of the ticket might be a bit misleading as to what the problem is. 

> it is intended behavior for a spinner to appear then disappear **once the response arrives with no results**.

**This** is the problem: the user doesn‘t know there were no results. They only ever see a spinner that shows up and disappears. Nothing else.

That is confusing. In my opinion, but also according to others who reported this as a possible bug („spinner shows up and disappears“).

(In an ideal state the user would know beforehand whether they can expect results or not (which is what the code intel badge is trying to give to the user). And we‘d also tell them (or help them to understand) for which tokens they could expect a token and for which ones they can‘t.)

So I think a first improvement could be to distinguish between „this is a valid token, but we have no results (cause no search-based results or no squirrel or no precise)“ and „we‘ll never have a result for this token, ever“ and show a „no results“ test in the hover for the former.

**Comment by mrnugget** [comment]
> UX/design question: when there is no result, should the hover popover tell the user?

cc @jjinnii here :)

**Comment by mrnugget** [comment]
More input to this: on a fresh instance, with no precise code intel, it's especially jarring to not get _anything_ when hovering over all the `glob.*` things here, but then getting this weirdly formatted `int                 len` hover tooltip.

https://user-images.githubusercontent.com/1185253/165925935-549f933f-71c3-4324-915d-a5f8204f99bb.mp4

**Comment by mrnugget** [comment]
Here's feedback from @vdavid on how confusing it is that sometimes a spinner shows up and then disappears: https://github.com/sourcegraph/sourcegraph/issues/32294#issuecomment-1113968694

**Comment by mrnugget** [comment]
cc @rrhyne This would be a great ticket for you to look at.

**Comment by rrhyne** [comment]
Could this be summarized by saying:  "We show a loader to indicate we are looking for something, not that we are loading data"?  I imagine the data itself is rather small, so loading it doesn't require a loading animation. 

If that's true, does anyone know why else we might show a loader? Perhaps @efritz? There is probably context I'm missing here, so be careful with this suggestion:

Don't show a loader. 
Do show clear error messages such as when the VPN logs someone out and the request takes a long time.

**Comment by efritz** [comment]
> I imagine the data itself is rather small, so loading it doesn't require a loading animation.

But the "loading" process is arbitrarily complex, including precise cross-repository resolution and searches (that might hit either an indexed or an unindexed path). Just because the actual data we send back is small doesn't mean there's not a lot going on under the hood that might make the user wait a bit.

**Comment by rrhyne** [comment]
Ah, I see. Still, I don't think we need to put such an attention-grabbing element up, only to remove it without telling the user what is occurring. There may be a solution that involves changing the token highlight or something else that is less obtrusive. 

Due to the context and exploration required I don't think this is something I should take on while only filling in, but I'd recommend it being considered an important design-debt issue.

**Comment by rrhyne** [comment]
It would be good to have some data on how often this disturbing pattern occurs and how often latency in returning the solution is high enough that we need to alert users that we're still working on it. 

@efritz do you know if this data is measured today? If not, I don't want to cause a request for on-prem telemetry as that can be time-consuming, but perhaps Sourcegraph.com could be instrumented?

**Comment by efritz** [comment]
@rrhyne Absolutely we can put some data in ... amplitude? Is that the right thing for Cloud today?

**Comment by rrhyne** [comment]
Yes, but all you need to do is fire an EventLog and it will get ingested automatically, no need to develop PINGs for it unless you want data from on-prem as well. 

I'm not sure how all the logic works so here are some thoughts on what to track:

* Create an event when code intel is not returned and the spinner is displayed: CodeIntelNotReturnedLoaderDisplayed

or

* Create a start hover event (if we don't have one) and fire a data returned event. I think we can look for the time between those in Amplitude. Not sure this is the best approach because it's going to generate a lot of event logs.

**Comment by mrnugget** [comment]
Just want to highlight this bit from the ticket itself here again:

> We only show the spinner after 600ms. That can lead to some rather tricky behaviour: if the first request is slow and doesn't find results, you'll see a spinner and then the hover disappears. If you then hover over the token again, the response from the previous request is cached, which means it'll not cross the 600ms, which means you won't see the spinner or a hover tooltip again.

This means we're talking about an edge case here, the tail end. So I'm wondering what we expect from the data: should it tell us that it happens often? Or should it confirm that it happens rarely?

I personally think that it's important even though it's rare, because it can be highly confusing.

And second thing: if we were to hide the loading spinner and simply never show it, only one part of the confusion would be fixed. The other part: sometimes we show a token, sometimes we don't and the user has no chance to recognize a pattern if they don't make a concious effort to look at (1) what code intel data do we have for the file I'm looking at (solved by Eric's new code intel badge) (2) can there be code intel for the token I'm hovering over (or is it a keyword, for example) (3) if the token should result in code intel, is it defined in _this_ repository for which I have data or is it defined in a dependency and do I have data for that?

**Comment by rrhyne** [comment]
Additional data would give us an idea of how important this is to fix. One form of data is the issues you've posted, but if this happens to 14% of users, even if it's rare, that should elevate its importance. If it rarely happens to 4%, that might be a different story.  If the team would already like to address this, then there is no need for the data.
